### PR TITLE
Allow creating annotated tags

### PR DESCRIPTION
**Pull request by ekohl**
When creating annotated tags (by passing a message into them), you can use "git describe".

### CODE REVIEW

File: tests.py [hunk 1]
```diff
@@ -686,6 +689,33 @@ def test_message_from_config_file(tmpdir, capsys, vcs):
...
+    tmpdir.join("VERSION").write("42.3.1")
+    check_call([vcs, "add", "VERSION"])
+    check_call([vcs, "commit", "-m", "initial commit"])
+
+    main(['patch', '--current-version', '42.3.1', '--commit', '--tag', 'VERSION', '--tag-name', 'ReleasedVersion-{new_version}'])
+
+    tag_out = check_output([vcs, {"git": "tag", "hg": "tags"}[vcs]])
+    assert b'ReleasedVersion-42.3.2' in tag_out
+
+    if vcs == "git":  # TODO: hg
```
Comment (ekohl) [inline tests.py]: I've revisited this and it appears there is no real ```git describe``` in mercurial so I don't think it's worth emulating.

**Comment by ekohl** [comment]
I reworked the tests so they also cover mercurial where appropriate. If you think it's alright then I'll update README as well.

**Comment by ekohl** [comment]
Btw, since I discovered that ```git describe --tags``` does what I want, my need for this has decreased. However, I do want to finish this.

**Comment by jgilchrist** [comment]
What's left to do on this before it can be merged?

**Comment by ekohl** [comment]
I think a rebase, a review and docs in README.

**Comment by ekohl** [comment]
Rebased

File: bumpversion/__init__.py [hunk 2]
```diff
@@ -146,8 +146,11 @@ def add_path(cls, path):
         subprocess.check_output(["git", "add", "--update", path])
 
     @classmethod
-    def tag(cls, name):
-        subprocess.check_output(["git", "tag", name])
+    def tag(cls, name, message):
+        command = ["git", "tag", name]
+        if message:
+            command += ['-m', message]
```
Comment (peritus) [inline bumpversion/__init__.py]: Should be using `--message`, because it's clearer.

File: bumpversion/__init__.py [hunk 3]
```diff
@@ -178,8 +181,11 @@ def add_path(cls, path):
         pass
 
     @classmethod
-    def tag(cls, name):
-        subprocess.check_output(["hg", "tag", name])
+    def tag(cls, name, message):
+        command = ["hg", "tag", name]
+        if message:
+            command += ['-m', message]
```
Comment (peritus) [inline bumpversion/__init__.py]: Should be using `--message`, because it's clearer.

**Comment by peritus** [comment]
I really like the _straightforwardness_ (is that even a word?) of this pull request, it just does the one thing it says on the box.

However, I'd really like "annotated commits" to be the default, that is, bumpversion should make annotated commits without users configuring that specifically because they know about annotated commits and they know that they're important.

That being said, `tag_message =` should have a nice default, similar to how `message =` (which would then be `commit_message =`, to avoid confusion) has a nice, recognizable default (compare https://github.com/peritus/bumpversion/blob/6ec908f018a827bfae94ed50baf241461eba7bcb/bumpversion/__init__.py#L784)

**Comment by peritus** [comment]
Also, what's common practice regarding information to include in annotated tags ? If there are any pointers to blog posts etc. describing what people put in them when releasing, now's the time.

**Comment by gst** [comment]
@peritus I agree with you  for the use of annotated tags by default.
For what they should include as information I would point you to this SO answer : http://stackoverflow.com/a/4971817/2451329

basically what they contains for the annotation itself, by default, is not that important, what's important is that with annotated tag you know who has submitted it and when it was done.
So I'd suggest simply put the same value as the tag value itself..

By the way : I'm waiting for this PR quite actively, I've a project where I work and we switched to gerrit code review and it requires annotated tags.. So for now I'm a bit stuck with current bumpversion ..

thanks.

**Comment by cmeeren** [comment]
@gst as a workaround you can tag manually (you can configure bumpversion not to tag the commits)

**Comment by ekohl** [comment]
I will have a look at rebasing + addressing the comments this weekend.

**Comment by gst** [comment]
@cmeeren thx for the tip, effectively, integrated in my .bumpversion.cfg :)

**Comment by ekohl** [comment]
@gst Thanks for the reminder :)

Updated. It now has a default tag message (can be disabled by passing an empty string) and passes ```--message``` to git/hg instead of ```-m```.

**Comment by ekohl** [comment]
@peritus mind having a look? I think it's ready now.

**Comment by ekohl** [comment]
@peritus ping

**Comment by dangra** [comment]
Please merge. It has been more than an year of development on this useful feature and looks more complete than ever.

**Comment by kynan** [comment]
@peritus Can this please be merged? I was about to implement this feature myself, glad I checked first :)

If the sticking point is docs I'll be happy to pr those once this is in.

**Comment by kynan** [comment]
Relates to #74

**Comment by schechter** [comment]
Can this feature be added please.  We running into issues because bumpversion is not making annotated tags.

**Comment by MarcMeszaros** [comment]
Any timeline on when this is being merged in? It would also help us.

**Comment by dbrgn** [comment]
Any news on this, now that Github actually shows signatures?

**Comment by nehalecky** [comment]
We really enjoy using bumpversion, but really need to have this feature added. When can we expect it?

**Comment by schechter** [comment]
@peritus  Is there anything additional you want in this PR?  I'm sure someone would do it if you let us know what you think it missing.

**Comment by anthrotype** [comment]
+1

**Comment by anthrotype** [comment]
Apologies for my insistence. I know there are more important things in life than annotated git tags.
However, this little change would make a lot of difference for my workflow.
I just noticed the last commit to `master` was more than a year ago. May I ask if this package is still maintained? I see it's being used by major projects, like setuptools for example...
Anyway, thanks for this great tool!

**Comment by jobevers** [comment]
+1 on annotated tags

**Comment by jaraco** [comment]
I'd like to see this feature as I use `bumpversion` for tagging setuptools releases, and without annotated tags, I have to remember to dual push (`git push` and `git push --tags`), whereas with annotated tags, I can rely on the [followTags option](http://stackoverflow.com/a/3745250/70170) to ensure that my tags are pushed.

**Comment by JamshedVesuna** [comment]
+1

**Comment by c4urself** [comment]
I was hoping that this would've been merged by now, unfortunately, it doesn't look like this repo is being actively maintained. In the interim, I've forked the project and named it `bump2version` (to reduce conflicts and allow a push to pypi). This version allows annotated tags support. I'm happy to change the README to point back to this repo if development resumes!

**Comment by anthrotype** [comment]
maybe we should consider proposing bumpversion to https://jazzband.co/

**Comment by ekohl** [comment]
For the lazy people like me a clickable link: https://github.com/c4urself/bump2version

@c4urself I considered forking myself but didn't want to commit myself to it and couldn't come up with a name. I like bump2version.

@anthrotype It may be a good but that still requires @peritus to transfer the repository.

**Comment by ulope** [comment]
@c4urself Awesome. Thanks for taking this on. Could you enable `Issues` in your fork?

**Comment by c4urself** [comment]
@ulope -- done
@ekohl -- want to join as a contributor?
@anthrotype -- i like that idea as well but contacted @peritus who suggested that a fork was the way to go for now

**Comment by ekohl** [comment]
@c4urself Done.

It would be nice if @peritus would officially endorse the fork, for example by making a note in the README both on github and pypi.

**Comment by ekohl** [comment]
Closing because bump2version has been available for quite a while.

**Pull request closed without merging**